### PR TITLE
mypy: Add annotations for bots.

### PR DESCRIPTION
**Pull request by skunkmb**
Add function annotations for 

 - *Incrementor* Bot, 
 - *Link Shortener* Bot, and 
 - *TicTacToe* Bot.

### CODE REVIEW

**Comment by neiljp** [comment]
@skunkmb I'll review the first two commits and will look at the third, but I'm currently working on refactoring it in #227, we might end up conflicting with each other there if my commits get accepted.

That said, if you'd like to take a look at that PR, there's nothing stopping you commenting on or reviewing it yourself, if you wanted :)

File: zulip_bots/zulip_bots/bots/incrementor/incrementor.py [hunk 1]
```diff
@@ -1,31 +1,33 @@
 # See readme.md for instructions on running this code.
 
+from zulip_bots.lib import ExternalBotHandler
```
Comment (neiljp) [inline zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: I'm not sure, but I believe the idea is to leave these as Any for now, pending a general interface class.
@showell ?

File: zulip_bots/zulip_bots/bots/incrementor/incrementor.py [hunk 2]
```diff
@@ -1,31 +1,33 @@
...
         storage = bot_handler.storage
         if not storage.contains('number') or not storage.contains('message_id'):
-            storage.put('number', 0)
-            storage.put('message_id', None)
+            # `storage.put` only accepts `str`s.
+            storage.put('number', str(0))
+            storage.put('message_id', '')
 
-    def handle_message(self, message, bot_handler):
+    def handle_message(self, message: dict, bot_handler: ExternalBotHandler) -> None:
```
Comment (neiljp) [inline zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: This should be `Dict`. You'll also need to specify what it is a `Dict` of, ie what other types.
These are elsewhere too.

File: zulip_bots/zulip_bots/bots/incrementor/incrementor.py [hunk 3]
```diff
@@ -1,31 +1,33 @@
...
         storage = bot_handler.storage
-        num = storage.get('number')
-        storage.put('number', num + 1)
-        if storage.get('message_id') is None:
+        num = int(storage.get('number'))
+        storage.put('number', str(num + 1))
```
Comment (neiljp) [inline zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: Does this not work already? Or is it a type issue?

File: zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py [hunk 4]
```diff
@@ -49,22 +53,22 @@ class TicTacToeGame(object):
                             "Here are the coordinates of each position. (Parentheses and spaces are optional). \n" \
                             "(1, 1)  (1, 2)  (1, 3) \n(2, 1)  (2, 2)  (2, 3) \n(3, 1) (3, 2) (3, 3) \n"
 
-    def __init__(self, board):
+    def __init__(self, board: list) -> None:
         self.board = board
 
-    def display_row(self, row):
+    def display_row(self, row: list) -> str:
```
Comment (neiljp) [inline zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py]: `list` -> `List`? Also elsewhere.

File: zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py [hunk 5]
```diff
@@ -115,7 +119,9 @@ def two_blanks(self, triplet, board):
             if len(blanks_list) == 2:
                 return blanks_list
 
-    def computer_move(self, board):
+        return None
```
Comment (neiljp) [inline zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py]: Returning `None` from a function returning `list` seems wrong?

File: zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py [hunk 6]
```diff
@@ -287,11 +293,17 @@ def handle_message(self, message, bot_handler):
         storage = bot_handler.storage
         if not storage.contains(original_sender):
             storage.put(original_sender, None)
+
+        # Some types are ignored later because `storage.get` is annotated to
+        # return a string only. However, it can actually return other types,
+        # like the board. For this reason, MyPy thinks `user_board`'s type is a
+        # `str` when it's actually a `list`, so `type: ignore` is used.
```
Comment (neiljp) [inline zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py]: I've not run into this before; I'll be interested to see if this remains an issue when you're using the capitalized list & dict forms.

File: zulip_bots/zulip_bots/bots/incrementor/incrementor.py (hunk 3 above)
Comment (skunkmb) [reply zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: As of now, `storage.get('number')` is annotated to return a `str` only, so trying to `+ 1` it causes an exception. Therefore, casting it to an `int` is needed.

I think using `# type: ignore` could also work, but I thought it was better to cast it than ignore the type entirely.

File: zulip_bots/zulip_bots/bots/incrementor/incrementor.py (hunk 1 above)
Comment (showell) [reply zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: yep, `Any` is good for now

File: zulip_bots/zulip_bots/bots/incrementor/incrementor.py (hunk 3 above)
Comment (showell) [reply zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: @skunkmb Ugh, the problem here is StateHandler.get/put are mis-annotated.  I'll fix this now.  I think your change is fine, though, as it's a bit more explicit.

Comment (showell) [reply zulip_bots/zulip_bots/bots/incrementor/incrementor.py]: fixed in https://github.com/zulip/python-zulip-api/commit/b4ff5230a9947aed4c3c2c88cf22f618233190ce

**Comment by showell** [comment]
@skunkmb I am going to merge the first two commits here.  You can link to this comment and tell the on-duty mentor that I am ok for them to approve this task.

@neiljp Since you're pretty heavily refactoring tictactoe, I'm going to let you do the annotation cleanup there at your leisure.

Thanks, guys!  (I'll close this once I merge from my laptop.)

**Comment by showell** [comment]
Hrm, I am also gonna let @neiljp fix `test_link_shortener.py` annotations as well.  This is unfortunately a fact of life in quick-moving projects--sometimes when two people are working on the same file, it's harder to resolve merge conflicts than simply re-do the work.  But hopefully this has been a good exercise, anyway (and I will merge the incrementor stuff).

**Comment by showell** [comment]
Here's the incrementor fix, after I made some of my own edits:

https://github.com/zulip/python-zulip-api/commit/3c66894aff10fc4a4868d755b5ad4ea473ba1563

Closing now.

**Comment by skunkmb** [comment]
@showell Thank you! 😄 You mentioned you were going to merge the first 2 commits (*Incrementor Bot* and *Link Shortener Bot*), but when I was looking at the commit history I only saw [one commit, merging the *Incrementor Bot* changes](https://github.com/zulip/python-zulip-api/commit/3c66894aff10fc4a4868d755b5ad4ea473ba1563):

recent commits

Are there still plans to merge the *Link Shortener Bot*'s annotation commit too? If not, I understand. Thanks!

**Comment by neiljp** [comment]
@skunkmb As mentioned above in the PR, @showell suggested I do the link_shortener work; I hope that's ok with you? Have you got credit for the GCI task ok?

I actually started by adding your repo as a remote, fetching it, then cherry-picking that commit onto my new branch; does that make sense? The new commit I added in 28687f1 straight away showed some corrections required from running `run-mypy`.

**Comment by skunkmb** [comment]
@neiljp OK, that's fine. Yes, I got credit for the task. Thanks!

**Pull request closed without merging**